### DISCUSSION TITLE
Properties with separate types when referenced internally and externally

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#4371] [Language Ideas]
I find myself coding properties like so:

```csharp
// private backing field that can be manipulated inside my class
private IList<string> names = new List<string>();

// expose it publicly but make it read only, so you can't go adding your own items or clearing it or whatever
public IEnumerable<string> Names => names;
```

It would be nice if there were some way to combine the backing field into the property somehow, so that from a `public` context (outside the class) the property appears to be `IEnumerable<string>`, but from a `private` context (inside the class) it appears to be `IList<string>`.

I'm not sure what the syntax for this would look like, maybe someone has an idea?

Of course the external type of a property would have to be a supertype of the property's internal type (or maybe an implicitly convertible type?), if they're not the same; it wouldn't make sense to declare a `string` property that appears to be an `int` from the outside!

**Comment by CyrusNajmabadi** [comment]
```c#
private readonly List<string> _names = new();
public IEnumerable<string> Names => _names;
```

Note: your pattern isn't safe here at all.  A consumer can still cast the result back to a List<string> and mutate it.

**Comment by HaloFour** [comment]
I don't think this use case warrants having a new form of auto-property.  I just don't think it's common enough.  Syntactically I don't think it makes sense to try to jam two members with different accessibility levels, different names and different types into a single member.  And auto-properties also don't allow the class to access the backing field directly, except in the one very specific case of read-only auto-properties initialized within the constructor.
